### PR TITLE
Added 'const' to a number of methods

### PR DESCRIPTION
**Pull request by gx740**
I hope this PR is final in which the qualifier `const` is added to methods.
In addition a lot of changes which improve readability.

### CODE REVIEW

File: Packet++/header/IPv4Layer.h [hunk 1]
```diff
@@ -228,18 +228,18 @@ namespace pcpp
...
 			if (m_Data == NULL)
 				return res;
 
-			size_t dataSize =  getDataSize();
+			size_t dataSize = getDataSize();
 			if (dataSize < 2)
 				return res;
 
-			uint8_t valueOffset = (uint8_t)(1);
+			uint8_t valueOffset = 1;
```
Comment (seladb) [inline Packet++/header/IPv4Layer.h]: I think removing the casting might cause a compilation warning on some platfroms

File: Packet++/header/IPv4Layer.h [hunk 2]
```diff
@@ -250,7 +250,7 @@ namespace pcpp
 
 				res.push_back(IPv4Address(curValue));
 
-				valueOffset += (uint8_t)(4);
+				valueOffset += 4;
```
Comment (seladb) [inline Packet++/header/IPv4Layer.h]: I think removing the casting might cause a compilation warning on some platfroms

File: Packet++/header/IPv4Layer.h [hunk 3]
```diff
@@ -280,7 +280,7 @@ namespace pcpp
 
 			res.type = (IPv4TimestampOptionValue::TimestampType)m_Data->recordValue[1];
 
-			uint8_t valueOffset = (uint8_t)(2);
+			uint8_t valueOffset = 2;
```
Comment (seladb) [inline Packet++/header/IPv4Layer.h]: I think removing the casting might cause a compilation warning on some platfroms

File: Packet++/header/IPv4Layer.h [hunk 4]
```diff
@@ -298,7 +298,7 @@ namespace pcpp
 				if (res.type == IPv4TimestampOptionValue::TimestampAndIP)
 					readIPAddr = !readIPAddr;
 
-				valueOffset += (uint8_t)(4);
+				valueOffset += 4;
```
Comment (seladb) [inline Packet++/header/IPv4Layer.h]: I think removing the casting might cause a compilation warning on some platfroms

File: Packet++/header/IPv4Layer.h [hunk 5]
```diff
@@ -580,7 +580,7 @@ namespace pcpp
 		/**
 		 * @return Size of IPv4 header (including IPv4 options if exist)
 		 */
-		size_t getHeaderLen() { return (size_t)(getIPv4Header()->internetHeaderLength*4) + m_TempHeaderExtension; }
+		size_t getHeaderLen() const { return (size_t)(getIPv4Header()->internetHeaderLength) * 4 + m_TempHeaderExtension; }
```
Comment (seladb) [inline Packet++/header/IPv4Layer.h]: The idea is to cast the while `getIPv4Header()->internetHeaderLength*4` expression to `size_t` to avoid compilation warning

File: Packet++/header/Layer.h [hunk 6]
```diff
@@ -124,12 +124,12 @@ namespace pcpp
...
 		 * @param[out] toArr The destination byte array
 		 */
-		void copyData(uint8_t* toArr);
+		void copyData(uint8_t* toArr) const;
 
 
 		// implement abstract methods
 
-		uint8_t* getDataPtr(size_t offset = 0) { return (uint8_t*)(m_Data + offset); }
+		uint8_t* getDataPtr(size_t offset = 0) const { return m_Data + offset; }
```
Comment (seladb) [inline Packet++/header/Layer.h]: this might cause compilation warning on some platforms

File: Packet++/src/DnsResource.cpp [hunk 7]
```diff
@@ -248,48 +248,30 @@ size_t DnsResource::getDataLength() const
...
 	case DNS_TYPE_MX:
-	{
-		return DnsResourceDataPtr(new MxDnsResourceData(resourceRawData, dataLength, this));
-		break;
-	}
+		return DnsResourceDataPtr(new MxDnsResourceData(resourceRawData, dataLength, const_cast<IDnsResource*>(static_cast<const IDnsResource*>(this))));
 
 	default:
-	{
 		return DnsResourceDataPtr(new GenericDnsResourceData(resourceRawData, dataLength));
```
Comment (seladb) [inline Packet++/src/DnsResource.cpp]: I think that adding the curly braces back will make the code look more readable

File: Packet++/src/GtpLayer.cpp [hunk 8]
```diff
@@ -587,22 +587,25 @@ void GtpV1Layer::parseNextLayer()
 
 	// GTP-U message, try to parse the next layer
 
-	uint8_t subProto = *(uint8_t*)(m_Data + headerLen);
+	uint8_t* payload = m_Data + headerLen;
```
Comment (seladb) [inline Packet++/src/GtpLayer.cpp]: this might cause compilation warning on some platforms (casting may be needed here)

File: Packet++/src/IPv6Extensions.cpp [hunk 9]
```diff
@@ -174,7 +174,7 @@ IPv6TLVOptionHeader::IPv6TLVOptionHeader(const std::vector<IPv6TLVOptionBuilder>
 	for (std::vector<IPv6TLVOptionBuilder>::const_iterator iter = options.begin(); iter != options.end(); iter++)
 	{
 		IPv6Option option = iter->build();
-		memcpy((uint8_t*)(getDataPtr() + offset), option.getRecordBasePtr(), option.getTotalSize());
+		memcpy(getDataPtr() + offset, option.getRecordBasePtr(), option.getTotalSize());
```
Comment (seladb) [inline Packet++/src/IPv6Extensions.cpp]: why did you remove the casting?

File: Packet++/src/IPv6Extensions.cpp [hunk 10]
```diff
@@ -207,15 +207,15 @@ IPv6RoutingHeader::IPv6RoutingHeader(uint8_t routingType, uint8_t segmentsLeft,
 
 	if (additionalRoutingDataLen > 0 && additionalRoutingData != NULL)
 	{
-		uint8_t* additionalDataPtr = (uint8_t*)(getDataPtr() + sizeof(ipv6_routing_header));
+		uint8_t* additionalDataPtr = getDataPtr() + sizeof(ipv6_routing_header);
```
Comment (seladb) [inline Packet++/src/IPv6Extensions.cpp]: why did you remove the casting?

File: Packet++/src/IPv6Extensions.cpp [hunk 11]
```diff
@@ -207,15 +207,15 @@ IPv6RoutingHeader::IPv6RoutingHeader(uint8_t routingType, uint8_t segmentsLeft,
...
 		memcpy(additionalDataPtr, additionalRoutingData, additionalRoutingDataLen);
 	}
 }
 
 uint8_t* IPv6RoutingHeader::getRoutingAdditionalData() const
 {
 	if (getExtensionLen() > sizeof(ipv6_routing_header))
-		return (uint8_t*)(getDataPtr() + sizeof(ipv6_routing_header));
+		return getDataPtr() + sizeof(ipv6_routing_header);
```
Comment (seladb) [inline Packet++/src/IPv6Extensions.cpp]: why did you remove the casting?

File: Packet++/src/IPv6Extensions.cpp [hunk 12]
```diff
@@ -263,15 +263,15 @@ IPv6AuthenticationHeader::IPv6AuthenticationHeader(uint32_t securityParametersIn
 
 	if (integrityCheckValueLen > 0 && integrityCheckValue != NULL)
 	{
-		uint8_t* icvPtr = (uint8_t*)(getDataPtr() + sizeof(ipv6_authentication_header));
+		uint8_t* icvPtr = getDataPtr() + sizeof(ipv6_authentication_header);
```
Comment (seladb) [inline Packet++/src/IPv6Extensions.cpp]: why did you remove the casting?

File: Packet++/src/IPv6Extensions.cpp [hunk 13]
```diff
@@ -263,15 +263,15 @@ IPv6AuthenticationHeader::IPv6AuthenticationHeader(uint32_t securityParametersIn
...
 		memcpy(icvPtr, integrityCheckValue, integrityCheckValueLen);
 	}
 }
 
 uint8_t* IPv6AuthenticationHeader::getIntegrityCheckValue() const
 {
 	if (getExtensionLen() > sizeof(ipv6_authentication_header))
-		return (uint8_t*)(getDataPtr() + sizeof(ipv6_authentication_header));
+		return getDataPtr() + sizeof(ipv6_authentication_header);
```
Comment (seladb) [inline Packet++/src/IPv6Extensions.cpp]: why did you remove the casting?

File: Packet++/header/IPv4Layer.h (hunk 1 above)
Comment (gx740) [reply Packet++/header/IPv4Layer.h]: It's just a positive constant. There is no any reason for warning. But I can roll back the changes.

I have looked into the build logs and not found any warnings that concern the changes made in this PR.

Comment (seladb) [reply Packet++/header/IPv4Layer.h]: I think I added that casting for a reason, probably a compilation warning. But it was a long time ago so I don't quite remember on which platform it was...

File: Packet++/header/IPv4Layer.h (hunk 5 above)
Comment (gx740) [reply Packet++/header/IPv4Layer.h]: I understand. I had moved the parenthesis to avoid following warning in Visual Studio:
`Warning	C26451	Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2)`

File: Packet++/header/IPv4Layer.h (hunk 1 above)
Comment (gx740) [reply Packet++/header/IPv4Layer.h]: Ok. I will return back the old code.

Comment (gx740) [reply Packet++/header/IPv4Layer.h]: Done

File: Packet++/header/IPv4Layer.h (hunk 5 above)
Comment (seladb) [reply Packet++/header/IPv4Layer.h]: Got it. But in the fix you've made, multiplying `(size_t)(getIPv4Header()->internetHeaderLength)` by 4 might make the compiler think the result will be larger that `size_t`, maybe we should add 2 castings there?

File: Packet++/header/Layer.h (hunk 6 above)
Comment (gx740) [reply Packet++/header/Layer.h]: The changes have been undone

File: Packet++/header/IPv4Layer.h (hunk 5 above)
Comment (gx740) [reply Packet++/header/IPv4Layer.h]: Like this?
`return (size_t)((uint16_t)(getIPv4Header()->internetHeaderLength) * 4) + m_TempHeaderExtension`

File: Packet++/src/DnsResource.cpp (hunk 7 above)
Comment (gx740) [reply Packet++/src/DnsResource.cpp]: The braces will be returned back.

File: Packet++/src/GtpLayer.cpp (hunk 8 above)
Comment (gx740) [reply Packet++/src/GtpLayer.cpp]: The casting has been added

File: Packet++/src/IPv6Extensions.cpp (hunk 9 above)
Comment (gx740) [reply Packet++/src/IPv6Extensions.cpp]: It is unnecessary because `getDataPtr()` returns `uint8_t*`

Comment (gx740) [reply Packet++/src/IPv6Extensions.cpp]: And `memcpy` gets a `void*` in the first parameter

File: Packet++/src/IPv6Extensions.cpp (hunk 10 above)
Comment (gx740) [reply Packet++/src/IPv6Extensions.cpp]: It is unnecessary because getDataPtr() returns uint8_t*

Comment (gx740) [reply Packet++/src/IPv6Extensions.cpp]: When a constant is added to a pointer the expression result will have the same type as pointer has.

File: Packet++/src/IPv6Extensions.cpp (hunk 11 above)
Comment (gx740) [reply Packet++/src/IPv6Extensions.cpp]: When a constant is added to a pointer the expression result will have the same type as pointer has. The `getDataPtr()` returns `uint8_t*`

File: Packet++/src/IPv6Extensions.cpp (hunk 12 above)
Comment (gx740) [reply Packet++/src/IPv6Extensions.cpp]: When a constant is added to a pointer the expression result will have the same type as pointer has.

File: Packet++/src/IPv6Extensions.cpp (hunk 13 above)
Comment (gx740) [reply Packet++/src/IPv6Extensions.cpp]: When a constant is added to a pointer the expression result will have the same type as pointer has.

File: Packet++/src/IPv6Extensions.cpp (hunk 9 above)
Comment (seladb) [reply Packet++/src/IPv6Extensions.cpp]: ok

File: Packet++/header/IPv4Layer.h (hunk 5 above)
Comment (seladb) [reply Packet++/header/IPv4Layer.h]: yes, that looks ok. Thanks!
